hazelcast/src/main/java/com/hazelcast/internal/probing/ProbingCycle.java

```diff
@@ -0,0 +1,222 @@
+     */
+
+    /**
+     * Resets the tags used by this cycle (to empty). This provides a convenient and
+     * format independent way of tagging the probed metrics.
+     * 
+     * After context (previously the prefix) is completed using
+     * {@link Tags#tag(CharSequence, CharSequence)} the {@code probe} methods are
+     * used to add measurements in this context. Then context is changed to next one
+     * and so forth.
+     * 
+     * @return the {@link Tags} abstraction to use to build the context for this
+     *         cycle.
+     */
+    Tags openContext();
+
+    /**
+     * The reason for {@link Tags} API is to allow "describing" complex "assembled"
+     * multi-part keys to the {@link ProbingCycle} without the need to create
+     * intermediate objects like {@link String} concatenation would.
+     */
+    interface Tags {
+
+        /**
+         * Append the tag to the context. If the name is identical to the last name the
+         * pair is not appended but the value just changes to the new value provided.
+         * This allows to setup tags that are identical for a collection first and just
+         * switch the varying tag to the current value in the loop. This does not work
+         * with more then one tag. In such cases the context has to be opened and
+         * reconstructed for each loop iteration.
+         * 
+         * The given {@link CharSequence} for {@code value} will always be escaped.
+         * 
+         * @param name not null, only guaranteed to be stable throughout the call
+         * @param value not null, only guaranteed to be stable throughout the call
+         * @return this {@link Tags} context for chaining
+         */
+        Tags tag(CharSequence name, CharSequence value);
+
+        /**
+         * This method is meat as a legacy support where keys are not build in terms of tags.
+         * It should only be used in exceptional cases. At some point we hopefully can remove it.
+         * 
+         * The given {@link CharSequence} will always be escaped.
+         * 
+         * @param s not null, only guaranteed to be stable throughout the call
+         * @return this {@link Tags} context for chaining
+         */
+        Tags append(CharSequence s);
+
+        /**
+         * Same as {@link #append(CharSequence)} + {@code append(".")} unless the prefix
+         * is empty in which case nothing is appended.
+         * 
+         * The given {@link CharSequence} will always be escaped.
+         * 
+         * @param prefix not null, only guaranteed to be stable throughout the call
+         * @return this {@link Tags} context for chaining
+         */
+        Tags prefix(CharSequence prefix);
+    }
+
+    /**
+     * Can be implemented by instances passed to {@link ProbingCycle#probe(Object)}
+     * (and its sibling methods) to provided the {@link Tags} context by the object
+     * itself instead of before calling {@code probe}.
+     */
+    interface Tagging {
```

Comment:
inner interface. Please pull it into top level.

Reply:
I think that makes it harder to understand and discover but as `Tagging` could be seen as something on top of base API you should have your top level type.